### PR TITLE
Use bounded channels in transport

### LINKED ISSUE #973: Replace unbounded channels with bounded ones
We have two usages of `mpsc::UnboudedSender`/`mpsc::UnboundedReceiver`. In order to avoid possible issues with the buffer growing forever, it would be nice to replace them with bounded channels (ie. `mpsc::Sender`/`mpsc::Receiver`).

### PR DESCRIPTION
**Pull request by mattrutherford**
Partially addresses issues #973. 

- [x] Replace unbounded channels in `transport::memory`
- [ ] ~~Replace unbounded channels in `nodes::handled_node_tasks`~~

To replace unbounded channels in `handled_node_tasks` would have implications that warrant further consideration, recommend to revisit at a later time.

### CODE REVIEW

File: core/src/transport/memory.rs [hunk 1]
```diff
@@ -29,19 +29,50 @@ use rw_stream_sink::RwStreamSink;
 use std::{collections::hash_map::Entry, error, fmt, io, num::NonZeroU64};
 
 lazy_static! {
-    static ref HUB: Mutex<FnvHashMap<NonZeroU64, mpsc::UnboundedSender<Channel<Bytes>>>> = Mutex::new(FnvHashMap::default());
+    static ref HUB: Mutex<FnvHashMap<NonZeroU64, mpsc::Sender<Channel<Bytes>>>> = Mutex::new(FnvHashMap::default());
 }
 
+static CHAN_BUF_SIZE: usize = 1000;
```
Comment (tomaka) [inline core/src/transport/memory.rs]: This doesn't need to be a constant. Inlining the value in the source code is enough. Also, 1000 is not a great value.

File: core/src/transport/memory.rs [hunk 2]
```diff
@@ -68,7 +99,7 @@ impl Transport for MemoryTransport {
 
         let actual_addr = Protocol::Memory(port.get()).into();
 
-        let (tx, rx) = mpsc::unbounded();
+        let (tx, rx) = mpsc::channel(CHAN_BUF_SIZE);
```
Comment (tomaka) [inline core/src/transport/memory.rs]: I think `2` is a good value. It's the number of maximum dialing connections that haven't been processed by the listener yet.
```suggestion
        let (tx, rx) = mpsc::channel(2);
```

File: core/src/transport/memory.rs [hunk 3]
```diff
@@ -94,20 +125,18 @@ impl Transport for MemoryTransport {
...
-            let (a_tx, a_rx) = mpsc::unbounded();
-            let (b_tx, b_rx) = mpsc::unbounded();
-            let a = RwStreamSink::new(Chan { incoming: a_rx, outgoing: b_tx });
-            let b = RwStreamSink::new(Chan { incoming: b_rx, outgoing: a_tx });
-            if tx.unbounded_send(b).is_err() {
-                return Err(TransportError::Other(MemoryTransportError::Unreachable));
-            }
-            a
+        if let Some(sender) = hub.get(&port) {
+            let (a_tx, a_rx) = mpsc::channel(CHAN_BUF_SIZE);
```
Comment (tomaka) [inline core/src/transport/memory.rs]: I'd say a value of `4096` for these two channels is good. It's the number of bytes that are pending.

File: core/src/transport/memory.rs [hunk 4]
```diff
@@ -29,19 +29,50 @@ use rw_stream_sink::RwStreamSink;
...
+            match self.sender.start_send(c) {
+                Err(_) => return Err(MemoryTransportError::Unreachable),
+                Ok(AsyncSink::NotReady(t)) => {
+                    self.channel_to_send = Some(t);
+                    return Ok(Async::NotReady)
+                },
+                _ => (),
+            }
+        }
+        match self.sender.poll_complete() {
```
Comment (tomaka) [inline core/src/transport/memory.rs]:
```suggestion
        match self.sender.close() {
```

`closed` automatically implies `poll_complete`.

File: core/src/transport/memory.rs [hunk 5]
```diff
@@ -29,19 +29,50 @@ use rw_stream_sink::RwStreamSink;
...
+            Err(_) => Err(MemoryTransportError::Unreachable),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Ok(Async::Ready(_)) => Ok(Async::Ready(self.channel_to_return.take().unwrap())),
```
Comment (tomaka) [inline core/src/transport/memory.rs]: We generally forbid `unwrap()` outside of tests in favour of a call to `expect()` that contains a message or a proof.
```suggestion
            Ok(Async::Ready(_)) => Ok(Async::Ready(self.channel_to_return.take()
                .expect("Future has been polled again after it is finished"))),
```

File: core/src/transport/memory.rs [hunk 6]
```diff
@@ -29,19 +29,50 @@ use rw_stream_sink::RwStreamSink;
...
+    static ref HUB: Mutex<FnvHashMap<NonZeroU64, mpsc::Sender<Channel<Bytes>>>> = Mutex::new(FnvHashMap::default());
 }
 
+static CHAN_BUF_SIZE: usize = 1000;
+
 /// Transport that supports `/memory/N` multiaddresses.
 #[derive(Debug, Copy, Clone, Default)]
 pub struct MemoryTransport;
 
+pub struct DialFuture {
```
Comment (tomaka) [inline core/src/transport/memory.rs]: Needs a small doc-comment. Nothing big, just *something*.
```suggestion
/// Connection to a `MemoryTransport` currently being opened.
pub struct DialFuture {
```

**Comment by mattrutherford** [comment]
Waiting for https://github.com/libp2p/rust-libp2p/pull/992 to get merged

**Pull request merged**